### PR TITLE
feat(learn): add word frequency challenge to rosetta section

### PR DESCRIPTION
**Pull request by moT01**
Made a PR in place of https://github.com/freeCodeCamp/freeCodeCamp/pull/38571 since the user was not responding and blocked edits from maintainers.

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] All the files I changed are in the same world language, for example: only English changes, or only Chinese changes, etc.

Closes #XXXXX

### CODE REVIEW

File: curriculum/challenges/english/10-coding-interview-prep/rosetta-code/word-frequency.english.md [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+## Tests
+<section id='tests'>
+
+``` yml
+tests:
+  - text: <code>wordFrequency</code> should be a function.
+    testString: assert(typeof wordFrequency == 'function');
+  - text: <code>wordFrequency</code> should return an array.
+    testString: assert(Array.isArray(wordFrequency("test")));
+  - text: <code>wordFrequency</code> should return the expected output for <code>("Hello hello world", 2)</code>
```
Comment (RandellDawson) [inline curriculum/challenges/english/10-coding-interview-prep/rosetta-code/word-frequency.english.md:32]:
```suggestion
  - text: <code>wordFrequency("Hello hello world", 2)</code> should return <code>[['hello', 2], ['world', 1]]</code>
```

File: curriculum/challenges/english/10-coding-interview-prep/rosetta-code/word-frequency.english.md [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    testString: assert.deepEqual(wordFrequency(example_1, 2), example_1_solution)
+  - text: <code>wordFrequency</code> should return the expected output for <code>("The quick brown fox jumped over the lazy dog", 4)</code>
```
Comment (RandellDawson) [inline curriculum/challenges/english/10-coding-interview-prep/rosetta-code/word-frequency.english.md:34]:
```suggestion
  - text: <code>wordFrequency("The quick brown fox jumped over the lazy dog", 4)</code> should return  <code>[['the', 2], ['brown', 1], ['dog', 1], ['fox', 1]]</code>
```

File: curriculum/challenges/english/10-coding-interview-prep/rosetta-code/word-frequency.english.md [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    testString: assert.deepEqual(wordFrequency(example_2, 4), example_2_solution)
+  - text: <code>wordFrequency</code> should return the expected output for <code>("Opensource opensource open-source open source", 1)</code>
```
Comment (RandellDawson) [inline curriculum/challenges/english/10-coding-interview-prep/rosetta-code/word-frequency.english.md:36]:
```suggestion
  - text: <code>wordFrequency("Opensource opensource open-source open source", 1)</code> should return 
 <code>[['opensource', 2]]</code>
```

File: curriculum/challenges/english/10-coding-interview-prep/rosetta-code/word-frequency.english.md [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    testString: assert.deepEqual(wordFrequency(example_3, 1), example_3_solution)
+  - text: <code>wordFrequency</code> should return the expected output for <code>("Apple application App apply applicative aPP aPPlE able", 3)</code>
```
Comment (RandellDawson) [inline curriculum/challenges/english/10-coding-interview-prep/rosetta-code/word-frequency.english.md:38]:
```suggestion
  - text: <code>wordFrequency("Apple application App apply applicative aPP aPPlE able", 3)</code> should return <code>[['app', 2], ['apple', 2], ['able', 1]]</code>
```

File: curriculum/challenges/english/10-coding-interview-prep/rosetta-code/word-frequency.english.md [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    testString: assert.deepEqual(wordFrequency(example_4, 3), example_4_solution)
+  - text: <code>wordFrequency</code> should return the expected output for <code>("c d a d c a b d d c", 4)</code>
```
Comment (RandellDawson) [inline curriculum/challenges/english/10-coding-interview-prep/rosetta-code/word-frequency.english.md:40]:
```suggestion
  - text: <code>wordFrequency("c d a d c a b d d c", 4)</code> should return <code>[['d', 4], ['c', 3], ['a', 2], ['b', 1]]</code>
```

File: curriculum/challenges/english/10-coding-interview-prep/rosetta-code/word-frequency.english.md [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    testString: assert.deepEqual(wordFrequency(example_5, 4), example_5_solution)
+  - text: <code>wordFrequency</code> should return an empty array for <code>("", 5)</code>
```
Comment (RandellDawson) [inline curriculum/challenges/english/10-coding-interview-prep/rosetta-code/word-frequency.english.md:42]:
```suggestion
  - text: <code>wordFrequency("", 5)</code> should return <code>[]</code>
```

**Review by RandellDawson — CHANGES_REQUESTED** [review]
@moT01 See my suggested changes.

Also, what should the following call return?
```
wordFrequency("c d a d c a b d c d c", 1)
```
I ask, because in this case, "c" and "d" would both have a frequency of `4`, so what determines which letter should be shown?

**Comment by thecodingaviator** [comment]
> Also, what should the following call return?
> 
> ```
> wordFrequency("c d a d c a b d c d c", 1)
> ```
> 
> I ask, because in this case, "c" and "d" would both have a frequency of `4`, so what determines which letter should be shown?

The Rosetta spec says "Words of equal frequency can be listed in any order."

**Comment by RandellDawson** [comment]
> The Rosetta spec says "Words of equal frequency can be listed in any order."

If that is the case @moT01, wouldn't we need to have two solutions for test 4, to allow both of the following?
```
[['app', 2], ['apple', 2], ['able', 1]]
```
and
```
[['apple', 2], ['app', 2], ['able', 1]]
```

**Comment by Sky020** [comment]
>  wouldn't we need to have two solutions for test 4, to allow both of the following?

@RandellDawson My understanding of `deepEqual` was that it allows for element order to be different.

**Comment by RandellDawson** [comment]
> @RandellDawson My understanding of deepEqual was that it allows for element order to be different.

@Sky020 If that is the case, then no extra solution should be necessary. Thanks for pointing that out.

**Comment by Sky020** [comment]
@RandellDawson Forgive me. I spoke too soon; after testing, I can confirm that `deepEqual` in Chaijs does care about element order. We would need to use `sameDeepMembers`

**Comment by RandellDawson** [comment]
> 'application' appears the same number of times as 'able'

@Sky020 I only see the word "application" once in the test code.  The other word is "applicative".

Right after your first reply, I tested and discovered (as you did) that `deepEqual` will not work.  Also, unless I am mistaken, using `sameDeepMembers` also only allows `[['app', 2], ['apple', 2], ['able', 1]]` to pass the 4th test.

**Comment by Sky020** [comment]
@RandellDawson , From what I have tested, `sameDeepMembers` passes with different order consistently:
image

I now see the issue with using it, though:
> return the n most common words in the file (and the number of their occurrences) **in decreasing frequency**.

So, we will need multiple solutions, as you mentioned in your original comment about this.

**Comment by moT01** [comment]
> Why is this not a valid answer:
> 
> [['apple', 2],['app', 2],['application', 1]];
> 

The way I understand it, that should be valid - meaning there could be a whole heck of a lot of solutions for that test. So you already mentioned that the first two words could be in any order, and then the last word could be any of the other four words that appear once. Making a total of 8 different solutions. I can try and fix it up - maybe make the sentences a little easier.

**Comment by moT01** [comment]
DeepEqual seems to be working for me - when I change the order of things, the tests don't pass. I must be missing something for why it doesn't work. Someone may have to explain it to me.

I simplified some of the tests so we don't have to check for multiple different solutions. If anyone has any better ideas, let me know.

**Comment by ojeytonwilliams** [comment]
Can't we just `.sort()` the two arrays we want to compare with `.deepEqual`? Then the tests would be more forgiving about implementation details.  If we go that way, I think we'd need to change the wording to reflect that.  i.e. `<code>wordFrequency("c d a d c a b d d c", 4)</code> should return <code>[['d', 4], ['c', 3], ['a', 2], ['b', 1]]</code>` -> `<code>wordFrequency("c d a d c a b d d c", 4)</code> should return an array containing <code>['d', 4], ['c', 3], ['a', 2], ['b', 1]</code>

**Comment by moT01** [comment]
The challenge states to return the array in a specific order, so I think sorting is out.

**Comment by ojeytonwilliams** [comment]
Oh, sorry, I skimmed.  What about `xs.sort((x,y) => y[1] - x[1] ? y[1] - x[1] : x[0].localeCompare(y[0]))` ? It should break ties on the order of the words.

**Comment by moT01** [comment]
Still not sure what's wrong with deepEqual - seemed to work when I tested it.

**Comment by Sky020** [comment]
@moT01 My testing revealed that, whilst this is an answer and passes the tests:
```
[['app', 2], ['apple', 2], ['able', 1]]
```
This should also pass the tests, but does not with `deepEqual`:
```
[['apple', 2], ['app', 2], ['able', 1]]
```

_provided it is tested against: `[['app', 2], ['apple', 2], ['able', 1]]`_

**Comment by moT01** [comment]
I changed a few of the tests so there's only one correct answer. Perhaps that wasn't what we wanted to do. Seemed like an easy solution.

**Comment by RandellDawson** [comment]
I think it should work regardless what tests we include or not include.  In fact, I think the test case I came up with should be included, since the instructions state that words with the same frequency can be any any order.  It is only one additional array to test against in the test, so I think that is the way to go.

**Comment by moT01** [comment]
I changed one of the challenges back to have multiple possible solutions. The only way I could figure out to test it was to stringify the array before comparing.

**Review by RandellDawson — DISMISSED** [review]

**Comment by RandellDawson** [comment]
@Sky020 I am happy with these latest changes that @moT01 made.  Merge if you approve.

File: curriculum/challenges/english/10-coding-interview-prep/rosetta-code/word-frequency.english.md [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+<section id='description'>
+Given a text string and an integer n, return the n most common words in the file (and the number of their occurrences) in decreasing frequency.
+</section>
+
+## Instructions
+
+<section id='instructions'>
+Write a function to count the occurrences of each word and return the n most commons words along with the number of their occurences in decreasing frequency.
+The function should return a 2D array with each of the elements in the following form: <code>[word, freq]</code>.  <code>word</code> should be the lowercase version of the word and <code>freq</code> the number denoting the count.
+Your function should be case insensitive, for example, the strings "Hello" and "hello" should be treated the same.
```
Comment (Sky020) [inline curriculum/challenges/english/10-coding-interview-prep/rosetta-code/word-frequency.english.md:20]:
```suggestion
The function should return an empty array, if no string is provided.
The function should be case insensitive, for example, the strings "Hello" and "hello" should be treated the same. 
```

**Review by ShaunSHamilton — CHANGES_REQUESTED** [review]
@moT01 I realise it is in the last test, but should we not add something to the instructions about returning an empty array?

Currently, the only mention (in the instructions) of what to return is:
> The function should return a 2D array...

Of which, `[]` is not.

**Comment by RandellDawson** [comment]
@Sky020 I am still happy with these changes.  Merge if you approve.

**Review by ShaunSHamilton — APPROVED** [review]
Thank you, for your contribution! 👍

We are happy to accept these changes and look forward to future contributions. 🎉

**Comment by Sky020** [comment]
@RandellDawson I am not allowed to merge, unless one more approving review gets made. So, you can do so, please.

**Review by RandellDawson — APPROVED** [review]

**Pull request merged**